R/credentials_gce.R

```diff
@@ -114,36 +138,112 @@ credentials_gce <- function(scopes = "https://www.googleapis.com/auth/cloud-plat
 #' @keywords internal
 #' @export
 GceToken <- R6::R6Class("GceToken", inherit = httr::Token2.0, list(
-  #' @description Print token
-  print = function(...) {
-    cat("<GceToken>")
+  #' @description Get an access for a GCE service account.
+  #' @param params A list of parameters for `fetch_gce_access_token()`.
+  #' @return A GceToken.
+  initialize = function(params) {
+    gargle_debug("GceToken initialize")
+    self$params <- params
+    self$init_credentials()
   },
-  #' @description Placeholder implementation of required method
+  #' @description Request an access token.
   init_credentials = function() {
-    self$credentials <- list(access_token = NULL)
+    gargle_debug("GceToken init_credentials")
+    token <- fetch_gce_access_token(
+      self$params$scope,
+      service_account = self$params$service_account
+    )
+
+    # find out the scopes actually obtained
+    # https://www.googleapis.com/oauth2/v3/tokeninfo
+    req <- request_build(
+      method = "GET",
+      path = "oauth2/v3/tokeninfo",
+      params = list(access_token = token$access_token),
+      base_url = "https://www.googleapis.com"
+    )
+    resp <- request_make(req)
+    info <- response_process(resp)
+    actual_scopes <- normalize_scopes(strsplit(info$scope, split = "\\s+")[[1]])
+
+    missing <- setdiff(self$params$scope, actual_scopes)
+    if (length(missing) > 0) {
+      gargle_debug(c(
+        "!" = "{cli::qty(length(missing))}{?This/These} requested \\
+             scope{?s} {?is/are} not among the scopes for the \\
+             access token returned by the metadata server:",
+        bulletize(missing, bullet = "x"),
+        "i" = "If there are problems downstream, this might be the root cause."
+      ))
```

Comment:
Another scope comparison block, this time comparing the token's *actual* scopes to those that were requested. It looks a lot like the previous one (comparing the instance scopes to requested scopes), but I've only copy/pasted twice, so no helper yet. I predict I'll revisit the GCE flow as we learn more from this extra info and I can tighten things up then.

Reply:
What this actually looks like in the case of `googledrive::drive_find()`:

```
GceToken initialize
GceToken init_credentials
! This requested scope is not among the scopes for the access token returned by the metadata server:
✖ https://www.googleapis.com/auth/drive
ℹ If there are problems downstream, this might be the root cause.
```